# 4.8 - Grouping Data & Aggregating Variables

In [2]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

In [3]:
# Creating a path:
path = r'D:\Nov Laptop\Ivan Dimitrov - Data Analyst (CF)\13-06-2023 Instacart Basket Analysis'

#### Step 1: Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [4]:
# Loading the file after the path
df_ords_prods_last_updated = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data','orders_products_merged_4.7_updated.pkl'))

In [5]:
# Importing the department file for reference
df_dept = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'))

In [6]:
df_dept

Unnamed: 0       department
0            1           frozen
1            2            other
2            3           bakery
3            4          produce
4            5          alcohol
5            6    international
6            7        beverages
7            8             pets
8            9  dry goods pasta
9           10             bulk
10          11    personal care
11          12     meat seafood
12          13           pantry
13          14        breakfast
14          15     canned goods
15          16       dairy eggs
16          17        household
17          18           babies
18          19           snacks
19          20             deli
20          21          missing

#### Step 2: In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [7]:
# Step 1: Check the file's dimensions
df_ords_prods_last_updated.shape

(32404859, 18)

In [8]:
# Step 2: Group by "product_name" 
df_ords_prods_last_updated.groupby('product_name')

In [9]:
# Step 3: Split the data into groups by "department_id" and then calculate the mean for the "order_number" column
df_ords_prods_last_updated.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### Step 3: Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

#### As we can see the results are quite different.
#### Subset orders minimal avg is (11.294069 - for dep 17 "household") - while for the Whole data set it's (15.215751 - for dep 5 "Alcohol")
#### Subset orders maximal avg is (19.463012 - for dep 16 "dairy eggs") - while for the Whole data set it's (22.902379 - for dep 21 "missing")

### Step 4: Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [10]:
# Step 1 of creating a loyalty flag is to create a "max_order" column
# to pile the aggregated results from grouping "user_id" and "order_number"
df_ords_prods_last_updated['max_order'] = df_ords_prods_last_updated.groupby(['user_id'])['order_number'].transform(np.max)

In [11]:
# Bypassing the limit of rows that can be shown
pd.options.display.max_rows = None

In [12]:
df_ords_prods_last_updated.head(100)

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

In [13]:
# Fixing the whole issue with "NaN" results:
df_ords_prods_last_updated['prices'].mean()

nan

In [14]:
df_ords_prods_last_updated['prices'] = df_ords_prods_last_updated['prices'].astype('float64')

In [15]:
df_ords_prods_last_updated['prices'].mean()

inf

In [16]:
inf_count = (df_ords_prods_last_updated['prices'] == float('inf')).sum()

In [17]:
inf_count

698

In [18]:
df_ords_prods_last_updated = df_ords_prods_last_updated[~(df_ords_prods_last_updated['prices'] == float('inf'))]

In [19]:
df_ords_prods_last_updated.reset_index(drop=True, inplace=True)

In [20]:
df_ords_prods_last_updated['prices'].mean()

9.825785349058837

In [21]:
# Create flags to assign "loyalty" labels to user ID based on max order value
df_ords_prods_last_updated.loc[df_ords_prods_last_updated['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

C:\Users\ivand\AppData\Local\Temp\ipykernel_15828\1618057219.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_ords_prods_last_updated.loc[df_ords_prods_last_updated['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'


In [22]:
df_ords_prods_last_updated.loc[(df_ords_prods_last_updated['max_order'] <= 40) & (df_ords_prods_last_updated['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [23]:
df_ords_prods_last_updated.loc[df_ords_prods_last_updated['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [24]:
# Check frequencies of loyalty flags for accuracy
df_ords_prods_last_updated['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876363
Loyal customer      10284010
New customer         6243788
Name: loyalty_flag, dtype: int64

### Step 5: The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [25]:
# To do so we create a group by "prices" 
df_ords_prods_last_updated.groupby('prices')

In [26]:
# Then we split the data into groups by "loyalty_flag" and then calculate the mean for each "prices" column
df_ords_prods_last_updated.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer     9.578717
New customer      10.059199
Regular customer   9.894029

### Step 6: The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:

#### Subestep: If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.” If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [27]:
# First we need to group our data by "user_id", generate average product price for each user,
# and create "avg_price" column for the aggregated results
df_ords_prods_last_updated['avg_price'] = df_ords_prods_last_updated.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\ivand\AppData\Local\Temp\ipykernel_15828\2960293675.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_ords_prods_last_updated['avg_price'] = df_ords_prods_last_updated.groupby(['user_id'])['prices'].transform(np.mean)


In [28]:
# Check grouping
df_ords_prods_last_updated.head(100)

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

In [29]:
# Create flags to assign "spending" labels to user ID based on average product price
df_ords_prods_last_updated.loc[df_ords_prods_last_updated['avg_price'] < 10, 'spending_flag'] = 'Low spender'

C:\Users\ivand\AppData\Local\Temp\ipykernel_15828\555147400.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_ords_prods_last_updated.loc[df_ords_prods_last_updated['avg_price'] < 10, 'spending_flag'] = 'Low spender'


In [30]:
df_ords_prods_last_updated.loc[df_ords_prods_last_updated['avg_price'] >= 10, 'spending_flag'] = 'High spender'

In [31]:
# Check frequencies of spending flags for accuracy
df_ords_prods_last_updated['spending_flag'].value_counts(dropna = False)

Low spender     31827629
High spender      576532
Name: spending_flag, dtype: int64

### Step 7: In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
#### If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
#### If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
#### If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [32]:
# To do so we need to group data by "user_id" column,
# generate median product of "days since prior order" column,
# and create "median_days" for aggregation results
df_ords_prods_last_updated['median_days'] = df_ords_prods_last_updated.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\ivand\AppData\Local\Temp\ipykernel_15828\715173083.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_ords_prods_last_updated['median_days'] = df_ords_prods_last_updated.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [33]:
# Check grouping
df_ords_prods_last_updated.head(10)

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   

  product_name  ...  _merge        price_label     busiest_day  \
0         Soda  ...    both  Mid-range product  Regularly busy   
1         Soda  ...    both  Mid-range product  Regularly busy   
2         Soda  ...    both  Mid-range product  Regularly busy   
3         Soda  ...    both  Mid-range product      Least busy   
4         Soda  ...    both  Mid-range product      Least busy   
5         Soda  ...    both  Mid-range product  Regularly busy   
6         Soda  ...    both  Mid-range product  Regularly busy   
7         Soda  ...    both  Mid-range product  Regularly busy   
8         Soda  ...    both  Mid-range product  Regularly busy   
9         Soda  ...    both  Mid-range product      Least busy   

     Busiest_Days busiest_period_of_day max_order  loyalty_flag avg_price  \
0  Regularly busy        Average Orders        10  New customer  6.367535   
1    Slowest days        Average Orders        10  New customer  6.367535   
2    Slowest days           Most Orders        10  New customer  6.367535   
3    Slowest days        Average Orders        10  New customer  6.367535   
4    Slowest days           Most Orders        10  New customer  6.367535   
5  Regularly busy        Average Orders        10  New customer  6.367535   
6    Busiest days           Most Orders        10  New customer  6.367535   
7    Busiest days           Most Orders        10  New customer  6.367535   
8    Busiest days           Most Orders        10  New customer  6.367535   
9    Slowest days        Average Orders        10  New customer  6.367535   

   spending_flag median_days  
0    Low spender        20.5  
1    Low spender        20.5  
2    Low spender        20.5  
3    Low spender        20.5  
4    Low spender        20.5  
5    Low spender        20.5  
6    Low spender        20.5  
7    Low spender        20.5  
8    Low spender        20.5  
9    Low spender        20.5  

[10 rows x 23 columns]

In [34]:
# Then we need to create flags and assign "freqency" labels to "user ID" based on the "median_days" since last order
df_ords_prods_last_updated.loc[df_ords_prods_last_updated['median_days'] > 20, 'frequency_flag'] = 'Non-frequent customer'

C:\Users\ivand\AppData\Local\Temp\ipykernel_15828\1580902318.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_ords_prods_last_updated.loc[df_ords_prods_last_updated['median_days'] > 20, 'frequency_flag'] = 'Non-frequent customer'


In [35]:
df_ords_prods_last_updated.loc[(df_ords_prods_last_updated['median_days'] > 10) & (df_ords_prods_last_updated['median_days'] <= 20), 'frequency_flag'] = 'Regular customer'

In [36]:
df_ords_prods_last_updated.loc[df_ords_prods_last_updated['median_days'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [37]:
# Check frequencies of spending flags for accuracy
df_ords_prods_last_updated['frequency_flag'].value_counts(dropna=False)

Frequent customer        21559380
Regular customer          7208433
Non-frequent customer     3636343
NaN                             5
Name: frequency_flag, dtype: int64

In [38]:
pd.options.display.max_rows = None

In [39]:
df_ords_prods_last_updated.head(100)

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

In [40]:
df_ords_prods_last_updated.loc[df_ords_prods_last_updated['prices'] >100, 'prices'] = np.nan

In [41]:
# Exporting data in pickle format since it's huge
df_ords_prods_last_updated.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'Orders_Products_Aggregated_4.8.pkl'))